File: abc288/c/main.c
#include <stdio.h>
#include <stdlib.h>

typedef struct {
	int *parent;
	int *size;
}	UF;

// Nは頂点数
UF *init_uf(int N) {
	UF *uf = calloc(1, sizeof(UF));
	uf->parent = calloc(N+1, sizeof(int));
	uf->size = calloc(N+1, sizeof(int));
	for (int i = 1; i <= N; i++) {
		uf->parent[i] = -1;
		uf->size[i] = 1;
	}
	return uf;
}
int	root(UF *uf, int x) {
	while(1) {
		if (uf->parent[x] == -1)
			break;
		x = uf->parent[x];
	}
	return x;
}
void	unite(UF *uf, int u, int v) {
	int RootU = root(uf, u);
	int RootV = root(uf, v);
	if (RootU == RootV)
		return ;
	if (uf->size[RootU] < uf->size[RootV]) {
		uf->parent[RootU] = RootV;
		uf->size[RootV] = uf->size[RootU] + uf->size[RootV];
	} else {
		uf->parent[RootV] = RootU;
		uf->size[RootU] = uf->size[RootV] + uf->size[RootU];
	}
}
int	uf_same(UF *uf, int u, int v) {
	return root(uf, u) == root(uf, v);
}

int	main(void)
{
	int n, m;

	scanf("%d %d", &n, &m);
	// printf("%d %d\n", n, m);

	UF *uf = init_uf(n);
	// printf("size=%d parent=%d\n", uf->size[1], uf->parent[1]);

	int ans = 0;
	for (int i = 1; i <= m; i++) {
		int a, b;
		scanf("%d %d", &a, &b);
		// printf("%d %d\n", a,b);
		
		if (uf_same(uf, a, b)){
			ans++;
		}
		unite(uf, a, b);
	}
	printf("%d\n", ans);
	return (0);
}

File: abc288/c/main.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"strconv"
)

type UnionFind struct {
	parent []int
	size   []int
}

// Nは頂点数
func NewUnionFind(n int) *UnionFind {
	uf := new(UnionFind)
	uf.parent = make([]int, n+1)
	uf.size = make([]int, n+1)
	for i := 1; i <= n; i++ {
		uf.parent[i] = -1
		uf.size[i] = 1
	}
	return uf
}
func (uf UnionFind) root(x int) int {
	for {
		if uf.parent[x] == -1 {
			break
		}
		x = uf.parent[x]
	}
	return x
}
func (uf UnionFind) unite(u, v int) {
	rootU := uf.root(u)
	rootV := uf.root(v)
	if rootU == rootV {
		return
	}
	if uf.size[rootU] < uf.size[rootV] {
		uf.parent[rootU] = rootV
		uf.size[rootV] = uf.size[rootU] + uf.size[rootV]
	} else {
		uf.parent[rootV] = rootU
		uf.size[rootU] = uf.size[rootV] + uf.size[rootU]
	}
}
func (uf UnionFind) same(u, v int) bool {
	return uf.root(u) == uf.root(v)
}

func main() {
	sc.Buffer(make([]byte, 128), 500000)
	sc.Split(bufio.ScanWords)
	n := scanInt()
	m := scanInt()
	// fmt.Printf("%d %d\n", n, m)

	ans := 0
	uf := NewUnionFind(n)
	for i := 1; i <= m; i++ {
		a, b := scanInt(), scanInt()

		// fmt.Printf("%d %d\n", a, b)
		if uf.same(a, b) {
			ans++
		}
		uf.unite(a, b)
	}
	fmt.Printf("%d\n", ans)
}

var sc = bufio.NewScanner(os.Stdin)

func scanInts(n int) []int {
	a := make([]int, n)
	for i := 0; i < n; i++ {
		a[i] = scanInt()
	}
	return a
}
func scanInt() int {
	sc.Scan()
	return atoi(sc.Text())
}
func atoi(nStr string) int {
	i, err := strconv.Atoi(nStr)
	if err != nil {
		panic(err)
	}
	return i
}
func scanText() string {
	sc.Scan()
	return sc.Text()
}
